### DISCUSSION TITLE
Frontend Development - related queries

### DISCUSSION THREAD

**Discussion by gsalunke** [fastn-stack/fastn#1137] [Q&A]
As discussed with @prafullsalunke taking down list of queries he had.

1. Is Module Federation supported?
2. Do we have an option for SSG (Static Site Generation)
3. Do we have context based router available in the stack (like NGINX Proxy does)
4. Can we dockerize the app for on premisis deployments?
5. For cloud deployment we have limitations on use of Cloudfront so can we host on our preferred environments.
6. Can we use node_modules inside the app and what are the other options for third-party s/w support?

@prafullsalunke you can add more queries in this discussion.

**Comment by amitu** [comment]
Thank you @prafullsalunke for trying us and the questions.

> 1. Is Module Federation supported?

[Module federation](https://webpack.js.org/concepts/module-federation/) is an ability to build multiple builds out of same code base. It is currently not supported by it is trivial to build. What is the use case?

> 2. Do we have an option for SSG (Static Site Generation)

Yes. `fastn build` can be used for that. This is the default mode we recommend right now for people to deploy.

> 3. Do we have context based router available in the stack (like NGINX Proxy does)

No, we route based on file system path, support [custom](https://fastn.com/custom-url/) and [dynamic](https://fastn.com/dynamic-urls/) urls.

> 4. Can we dockerize the app for on premisis deployments?

We offer a [single binary](https://github.com/fastn-stack/fastn/releases) you can install anywhere. Creating a Dockerfile should be trivial:

```Dockerfile
FROM node:18-alpine
RUN curl -fsSL https://fastn.com/install.sh | bash
CMD ["fastn", "serve"]
EXPOSE 8000
```

Not tested, but it should be this easy.

> 5. For cloud deployment we have limitations on use of Cloudfront so can we host on our preferred environments.

If you are generating static files using `fastn build`, you can easily upload them to a bucket in S3 and point your Cloudfront to the bucket. 

> 6. Can we use node_modules inside the app and what are the other options for third-party s/w support?

`fastn` has nothing to do with what else you have in the package. `fastn build` may become slow if it ends up processing files in node_modules, as by default it process files in current directly. We do have `.gitignore` handling, so ideally if you have used `git` ignored it it will not be a problem. 

It's just a binary that converts `.ftd` file to `.html` etc. Ideally do not pollute your fastn package folder with node stuff, you can use node from another folder.

If you are talking about using our  [`js` integration](https://fastn.com/use-js/), you can use whatever to build your JS file, `fastn` just needs the location of the final JS file you want included in the HTML.

Do let me know if you have any follow up questions.
